go/saltpack/decrypt.go

```diff
@@ -257,13 +258,19 @@ func (ds *decryptStream) processEncryptionBlock(bl *EncryptionBlock) ([]byte, er
 	}
 
 	nonce := ds.nonce.ForPayloadBox(blockNum)
+	ciphertext := bl.PayloadCiphertext
+	hash := sha512.Sum512(ciphertext)
 
-	tag, err := ds.tagKey.Unbox(nonce, bl.TagCiphertexts[ds.position])
+	hashBox, err := ds.tagKey.Box(nonce, hash[:])
 	if err != nil {
+		return nil, err
+	}
+	ourAuthenticator := hashBox[:secretbox.Overhead]
+
+	if !hmac.Equal(ourAuthenticator, bl.HashAuthenticators[ds.position]) {
```

Comment:
I thought about calling it `BadHashAuthenticator`, but then I decided `BadTag` was still valid. What's changed is the input the tag is computed over.